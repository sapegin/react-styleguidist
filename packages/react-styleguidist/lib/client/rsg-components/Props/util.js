import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.array.iterator";
import "core-js/modules/es6.object.keys";
import "core-js/modules/es6.function.name";
import "core-js/modules/es6.regexp.replace";

function _objectSpread(target) { for (let i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; let ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Remove quotes around given string.
 *
 * @param {string} string
 * @returns {string}
 */
export function unquote(string) {
  return string.replace(/^['"]|['"]$/g, '');
}
/**
 * Return prop type object.
 *
 * @param {object} prop
 * @returns {object}
 */

export function getType(prop) {
  if (prop.flowType) {
    if (prop.flowType.name === 'union' && prop.flowType.elements.every(function (elem) {
      return elem.name === 'literal';
    })) {
      return _objectSpread({}, prop.flowType, {
        name: 'enum',
        value: prop.flowType.elements
      });
    }

    return prop.flowType;
  }

  return prop.type;
}
/**
 * Show starting and ending whitespace around given string.
 *
 * @param {string} string
 * @returns {string}
 */

export function showSpaces(string) {
  return string.replace(/^\s|\s$/g, '␣');
}