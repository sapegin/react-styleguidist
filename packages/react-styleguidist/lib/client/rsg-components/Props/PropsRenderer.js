import "core-js/modules/es6.object.assign";
import "core-js/modules/es6.array.from";
import "core-js/modules/es6.regexp.to-string";
import "core-js/modules/es7.symbol.async-iterator";
import "core-js/modules/es6.symbol";
import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.function.name";

import React from 'react';
import PropTypes from 'prop-types';
import Group from 'react-group';
import objectToString from 'javascript-stringify';
import Arguments from 'rsg-components/Arguments';
import Argument from 'rsg-components/Argument';
import Code from 'rsg-components/Code';
import JsDoc from 'rsg-components/JsDoc';
import Markdown from 'rsg-components/Markdown';
import Name from 'rsg-components/Name';
import Type from 'rsg-components/Type';
import Text from 'rsg-components/Text';
import Para from 'rsg-components/Para';
import Table from 'rsg-components/Table';
import { unquote, getType, showSpaces } from './util';

function _extends() { _extends = Object.assign || function (target) { for (let i = 1; i < arguments.length; i++) { const source = arguments[i]; for (const key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") {return Array.from(iter);} }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function renderType(type) {
  if (!type) {
    return 'unknown';
  }

  const name = type.name;

  switch (name) {
    case 'arrayOf':
      return "".concat(type.value.name, "[]");

    case 'objectOf':
      return "{".concat(renderType(type.value), "}");

    case 'instanceOf':
      return type.value;

    default:
      return name;
  }
}

function renderFlowType(type) {
  if (!type) {
    return 'unknown';
  }

  const name = type.name;

      
const raw = type.raw;

      
const value = type.value;

  switch (name) {
    case 'enum':
      return name;

    case 'literal':
      return value;

    case 'signature':
      return renderComplexType(type.type, raw);

    case 'union':
    case 'tuple':
      return renderComplexType(name, raw);

    default:
      return raw || name;
  }
}

function renderComplexType(name, title) {
  return React.createElement(Text, {
    size: "small",
    underlined: true,
    title
  }, React.createElement(Type, null, name));
}

function renderEnum(prop) {
  if (!Array.isArray(getType(prop).value)) {
    return React.createElement("span", null, getType(prop).value);
  }

  const values = getType(prop).value.map(function (_ref) {
    const value = _ref.value;
    return React.createElement(Code, {
      key: value
    }, showSpaces(unquote(value)));
  });
  return React.createElement("span", null, "One of:", ' ', React.createElement(Group, {
    separator: ", ",
    inline: true
  }, values));
}

function renderShape(props) {
  const rows = [];

  for (const name in props) {
    const prop = props[name];
    const defaultValue = renderDefault(prop);
    const description = prop.description;
    rows.push(React.createElement("div", {
      key: name
    }, React.createElement(Name, null, name), ': ', React.createElement(Type, null, renderType(prop)), defaultValue && ' — ', defaultValue, description && ' — ', description && React.createElement(Markdown, {
      text: description,
      inline: true
    })));
  }

  return rows;
}

const defaultValueBlacklist = ['null', 'undefined'];

function renderDefault(prop) {
  // Workaround for issue https://github.com/reactjs/react-docgen/issues/221
  // If prop has defaultValue it can not be required
  if (prop.defaultValue) {
    if (prop.type || prop.flowType) {
      const propName = prop.type ? prop.type.name : prop.flowType.type;

      if (defaultValueBlacklist.indexOf(prop.defaultValue.value) > -1) {
        return React.createElement(Code, null, showSpaces(unquote(prop.defaultValue.value)));
      } else if (propName === 'func' || propName === 'function') {
        return React.createElement(Text, {
          size: "small",
          color: "light",
          underlined: true,
          title: showSpaces(unquote(prop.defaultValue.value))
        }, "Function");
      } else if (propName === 'shape' || propName === 'object') {
        try {
          // We eval source code to be able to format the defaultProp here. This
          // can be considered safe, as it is the source code that is evaled,
          // which is from a known source and safe by default
          // eslint-disable-next-line no-eval
          const object = eval("(".concat(prop.defaultValue.value, ")"));
          return React.createElement(Text, {
            size: "small",
            color: "light",
            underlined: true,
            title: objectToString(object, null, 2)
          }, "Shape");
        } catch (e) {
          // eval will throw if it contains a reference to a property not in the
          // local scope. To avoid any breakage we fall back to rendering the
          // prop without any formatting
          return React.createElement(Text, {
            size: "small",
            color: "light",
            underlined: true,
            title: prop.defaultValue.value
          }, "Shape");
        }
      }
    }

    return React.createElement(Code, null, showSpaces(unquote(prop.defaultValue.value)));
  } else if (prop.required) {
    return React.createElement(Text, {
      size: "small",
      color: "light"
    }, "Required");
  }

  return '';
}

function renderDescription(prop) {
  const description = prop.description;

      
const _prop$tags = prop.tags;

      
const tags = _prop$tags === void 0 ? {} : _prop$tags;
  const extra = renderExtra(prop);

  const args = _toConsumableArray(tags.arg || []).concat(_toConsumableArray(tags.argument || []), _toConsumableArray(tags.param || []));

  const returnDocumentation = tags.return && tags.return[0] || tags.returns && tags.returns[0];
  return React.createElement("div", null, description && React.createElement(Markdown, {
    text: description
  }), extra && React.createElement(Para, null, extra), React.createElement(JsDoc, tags), args.length > 0 && React.createElement(Arguments, {
    args,
    heading: true
  }), returnDocumentation && React.createElement(Argument, _extends({}, returnDocumentation, {
    returns: true
  })));
}

function renderExtra(prop) {
  const type = getType(prop);

  if (!type) {
    return null;
  }

  switch (type.name) {
    case 'enum':
      return renderEnum(prop);

    case 'union':
      return renderUnion(prop);

    case 'shape':
      return renderShape(prop.type.value);

    case 'arrayOf':
      if (type.value.name === 'shape') {
        return renderShape(prop.type.value.value);
      }

      return null;

    case 'objectOf':
      if (type.value.name === 'shape') {
        return renderShape(prop.type.value.value);
      }

      return null;

    default:
      return null;
  }
}

function renderUnion(prop) {
  const type = getType(prop);

  if (!Array.isArray(type.value)) {
    return React.createElement("span", null, type.value);
  }

  const values = type.value.map(function (value, index) {
    return React.createElement(Type, {
      key: "".concat(value.name, "-").concat(index)
    }, renderType(value));
  });
  return React.createElement("span", null, "One of type:", ' ', React.createElement(Group, {
    separator: ", ",
    inline: true
  }, values));
}

function renderName(prop) {
  const name = prop.name;

      
const _prop$tags2 = prop.tags;

      
const tags = _prop$tags2 === void 0 ? {} : _prop$tags2;
  return React.createElement(Name, {
    deprecated: !!tags.deprecated
  }, name);
}

function renderTypeColumn(prop) {
  if (prop.flowType) {
    return React.createElement(Type, null, renderFlowType(getType(prop)));
  }

  return React.createElement(Type, null, renderType(getType(prop)));
}

export function getRowKey(row) {
  return row.name;
}
export var columns = [{
  caption: 'Prop name',
  render: renderName
}, {
  caption: 'Type',
  render: renderTypeColumn
}, {
  caption: 'Default',
  render: renderDefault
}, {
  caption: 'Description',
  render: renderDescription
}];
export default function PropsRenderer(_ref2) {
  const props = _ref2.props;
  return React.createElement(Table, {
    columns,
    rows: props,
    getRowKey
  });
}
PropsRenderer.propTypes = {
  props: PropTypes.array.isRequired
};