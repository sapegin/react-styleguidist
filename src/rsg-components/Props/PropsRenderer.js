import React, { PropTypes } from 'react';
import Code from 'rsg-components/Code';
import Markdown from 'rsg-components/Markdown';
import Styled from 'rsg-components/Styled';
import Group from 'react-group';
import { unquote, getType, showSpaces } from './util';

const styles = ({ font, border, light, lightest, name, type }) => ({
	table: {
		width: '100%',
		borderCollapse: 'collapse',
	},
	tableHead: {
		borderBottom: [[1, border, 'solid']],
	},
	cell: {
		paddingRight: 15,
		paddingTop: 6,
		verticalAlign: 'top',
		fontFamily: font,
		fontSize: 13,
	},
	cellHeading: {
		paddingRight: 15,
		paddingBottom: 6,
		textAlign: 'left',
		fontFamily: font,
		fontWeight: 'bold',
		fontSize: 13,
	},
	cellDesc: {
		width: '99%',
		paddingLeft: 15,
	},
	required: {
		fontFamily: font,
		fontSize: 13,
		color: light,
	},
	name: {
		fontSize: 13,
		color: name,
	},
	type: {
		fontSize: 13,
		color: type,
	},
	function: {
		fontFamily: font,
		fontSize: 13,
		color: light,
		borderBottom: `1px dotted ${lightest}`,
	},
});

function renderType(type) {
	if (!type) {
		return 'unknown';
	}

	const { name } = type;

	switch (name) {
		case 'arrayOf':
			return `${type.value.name}[]`;
		case 'objectOf':
			return `{${renderType(type.value)}}`;
		case 'instanceOf':
			return type.value;
		default:
			return name;
	}
}

function renderEnum(prop) {
	if (!Array.isArray(getType(prop).value)) {
		return <span>{getType(prop).value}</span>;
	}

	const values = getType(prop).value.map(({ value }) => (
		<Code key={value}>{showSpaces(unquote(value))}</Code>
	));
	return (
		<span>One of: <Group separator=", " inline>{values}</Group></span>
	);
}

export function PropsRenderer({ classes, props }) {
	function renderRows(props) {
		const rows = [];
		for (const name in props) {
			const prop = props[name];
			rows.push(
				<tr key={name}>
					<td className={classes.cell}><Code className={classes.name}>{name}</Code></td>
					<td className={classes.cell}><Code className={classes.type}>{renderType(getType(prop))}</Code></td>
					<td className={classes.cell}>{renderDefault(prop)}</td>
					<td className={classes.cell + ' ' + classes.cellDesc}>{renderDescription(prop)}</td>
				</tr>
			);
		}
		return rows;
	}

	function renderDefault(prop) {
		if (prop.required) {
			return (
				<span className={classes.required}>Required</span>
			);
		}
		else if (prop.defaultValue) {
			if (prop.type && prop.type.name === 'func') {
				return (
					<span className={classes.function} title={showSpaces(unquote(prop.defaultValue.value))}>Function</span>
				);
			}

			return (
				<Code>{showSpaces(unquote(prop.defaultValue.value))}</Code>
			);
		}
		return '';
	}

	function renderDescription(prop) {
		const { description } = prop;
		const extra = renderExtra(prop);
		return (
			<Group separator={<br />}>
				{description && <Markdown text={description} inline />}
				{extra}
			</Group>
		);
	}

	function renderExtra(prop) {
		const type = getType(prop);

		if (!type) {
			return null;
		}

		switch (type.name) {
			case 'enum':
				return renderEnum(prop);
			case 'union':
				return renderUnion(prop);
			case 'shape':
				return renderShape(prop.type.value);
			case 'arrayOf':
				if (type.value.name === 'shape') {
					return renderShape(prop.type.value.value);
				}
				return null;
			case 'objectOf':
				if (type.value.name === 'shape') {
					return renderShape(prop.type.value.value);
				}
				return null;
			default:
				return null;
		}
	}

	function renderUnion(prop) {
		if (!Array.isArray(getType(prop).value)) {
			return <span>{getType(prop).value}</span>;
		}

		const values = getType(prop).value.map(value => (
			<Code key={value.name} className={classes.type}>{renderType(value)}</Code>
		));
		return (
			<span>One of type: <Group separator=", " inline>{values}</Group></span>
		);
	}

	function renderShape(props) {
		const rows = [];
		for (const name in props) {
			const prop = props[name];
			const defaultValue = renderDefault(prop);
			const description = prop.description;
			rows.push(
				<div key={name}>
					<Code className={classes.name}>{name}</Code>{': '}
					<Code className={classes.type}>{renderType(prop)}</Code>
					{defaultValue && ' — '}{defaultValue}
					{description && ' — '}
					{description && <Markdown text={description} inline />}
				</div>
			);
		}
		return rows;
	}

	return (
		<table className={classes.table}>
			<thead className={classes.tableHead}>
				<tr>
					<th className={classes.cellHeading}>Name</th>
					<th className={classes.cellHeading}>Type</th>
					<th className={classes.cellHeading}>Default</th>
					<th className={classes.cellHeading + ' ' + classes.cellDesc}>Description</th>
				</tr>
			</thead>
			<tbody>
				{renderRows(props)}
			</tbody>
		</table>
	);
}

PropsRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	props: PropTypes.object.isRequired,
};

export default Styled(styles)(PropsRenderer);
