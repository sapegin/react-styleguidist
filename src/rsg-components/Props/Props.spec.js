import React from 'react';
import { parse } from 'react-docgen';
import PropsRenderer, { propsToArray, columns } from './PropsRenderer';
import { unquote, getType, showSpaces } from './util';

// Test renderers with clean readable snapshot diffs
// eslint-disable-next-line react/prop-types
export default function ColumnsRenderer({ props }) {
	return (
		<ul>
			{propsToArray(props).map((row, rowIdx) =>
				<li key={rowIdx}>
					{columns.map(({ render }, colIdx) =>
						<div key={colIdx}>
							{render(row)}
						</div>
					)}
				</li>
			)}
		</ul>
	);
}

function render(propTypes, defaultProps = []) {
	const props = parse(`
		import { Component } from 'react';
		import PropTypes from 'prop-types';
		export default class Cmpnt extends Component {
			static propTypes = {
				${propTypes.join(',')}
			}
			static defaultProps = {
				${defaultProps.join(',')}
			}
			render() {
			}
		}
	`);
	return shallow(<ColumnsRenderer props={props.props} />);
}

describe('PropsRenderer', () => {
	it('should render a table', () => {
		const actual = shallow(
			<PropsRenderer
				props={{ color: { type: { name: 'string' }, required: false, description: '' } }}
			/>
		);

		expect(actual).toMatchSnapshot();
	});
});

describe('props columns', () => {
	it('should render PropTypes.string', () => {
		const actual = render(['color: PropTypes.string']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.string with a default value', () => {
		const actual = render(['color: PropTypes.string'], ['color: "pink"']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.string.isRequired', () => {
		const actual = render(['color: PropTypes.string.isRequired']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.arrayOf', () => {
		const actual = render(['colors: PropTypes.arrayOf(PropTypes.string)']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.arrayOf(PropTypes.shape)', () => {
		const actual = render([
			'foos: PropTypes.arrayOf(PropTypes.shape({bar: PropTypes.number, baz: PropTypes.any}))',
		]);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.instanceOf', () => {
		const actual = render(['num: PropTypes.instanceOf(Number)']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.shape', () => {
		const actual = render([
			'foo: PropTypes.shape({bar: PropTypes.number.isRequired, baz: PropTypes.any})',
		]);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.shape with formatted defaultProps', () => {
		const actual = render(
			[
				`
				foo: PropTypes.shape({
					bar: PropTypes.number.isRequired,
					baz: PropTypes.any,
				})
			`,
			],
			[
				`
				foo: {
					bar: 123, baz() {
						return 'foo';
					},
					bing() {
						return 'badaboom';
					},
					trotskij: () => 1935,
					qwarc: { si: 'señor', },
				}
			`,
			]
		);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.shape with description', () => {
		const actual = render([
			`foo: PropTypes.shape({
			/**
			 * Number
			 */
			bar: PropTypes.number.isRequired,
			/** Any */
			baz: PropTypes.any
		})`,
		]);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.objectOf', () => {
		const actual = render(['colors: PropTypes.objectOf(PropTypes.string)']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.objectOf(PropTypes.shape)', () => {
		const actual = render([
			`colors: PropTypes.objectOf(
			PropTypes.shape({
				bar: PropTypes.number.isRequired,
				baz: PropTypes.any
			})
		)`,
		]);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.oneOf', () => {
		const actual = render(['size: PropTypes.oneOf(["small", "normal", "large"])']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.oneOfType', () => {
		const actual = render(['union: PropTypes.oneOfType([PropTypes.string, PropTypes.number])']);

		expect(actual).toMatchSnapshot();
	});

	it('should render description in Markdown', () => {
		const actual = render(['/**\n * Label\n */\ncolor: PropTypes.string']);

		expect(actual).toMatchSnapshot();
	});

	it('should render unknown proptype for a prop when a relevant proptype is not assigned', () => {
		const actual = render([], ['color: "pink"']);

		expect(actual).toMatchSnapshot();
	});

	it('should render function body in tooltip', () => {
		const actual = render(['fn: PropTypes.func'], ['fn: (e) => console.log(e)']);

		expect(actual).toMatchSnapshot();
	});

	it('should render arguments from JsDoc tags', () => {
		const props = {
			size: {
				type: {
					name: 'number',
				},
				required: false,
				description: 'Test description',
				tags: {
					arg: [
						{
							name: 'Foo',
							description: 'Converts foo to bar',
							type: { name: 'Array' },
						},
					],
					param: [
						{
							name: 'Bar',
						},
					],
				},
			},
		};
		const actual = shallow(<ColumnsRenderer props={props} />);

		expect(actual).toMatchSnapshot();
	});

	it('should render name as deprecated when tag deprecated is present', () => {
		const props = {
			size: {
				type: {
					name: 'number',
				},
				required: false,
				description: 'Test description',
				tags: {
					deprecated: [
						{
							title: 'deprecated',
							description: 'Do not use.',
						},
					],
				},
			},
		};
		const actual = shallow(<ColumnsRenderer props={props} />);

		expect(actual).toMatchSnapshot();
	});
});

describe('unquote', () => {
	it('should remove double quotes around the string', () => {
		const result = unquote('"foo"');
		expect(result).toBe('foo');
	});

	it('should remove single quotes around the string', () => {
		const result = unquote("'foo'");
		expect(result).toBe('foo');
	});

	it('should not remove quotes in the middle of the string', () => {
		const result = unquote('foo"bar');
		expect(result).toBe('foo"bar');
	});
});

describe('getType', () => {
	it('should return .type or .flowType property', () => {
		const result = getType({
			type: 'foo',
			flowType: 'bar',
		});
		expect(result).toBe('bar');
	});
});

describe('showSpaces', () => {
	it('should replace leading and trailing spaces with a visible character', () => {
		const result = showSpaces(' pizza ');
		expect(result).toBe('␣pizza␣');
	});
});
