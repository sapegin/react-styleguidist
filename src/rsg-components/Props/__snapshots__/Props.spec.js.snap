// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropsRenderer should render a table 1`] = `
<Styled(Table)
  columns={
    Array [
      Object {
        "caption": "Prop name",
        "render": [Function],
      },
      Object {
        "caption": "Type",
        "render": [Function],
      },
      Object {
        "caption": "Default",
        "render": [Function],
      },
      Object {
        "caption": "Description",
        "render": [Function],
      },
    ]
  }
  getRowKey={[Function]}
  rows={
    Array [
      Object {
        "description": "",
        "name": "color",
        "required": false,
        "type": Object {
          "name": "string",
        },
      },
    ]
  }
/>
`;

exports[`props columns should render PropTypes.arrayOf 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        colors
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        string[]
      </Styled(Type)>
    </div>
    <div />
    <div>
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.arrayOf(PropTypes.shape) 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        foos
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        shape[]
      </Styled(Type)>
    </div>
    <div />
    <div>
      <div>
        <Styled(Para)>
          <div>
            <Styled(Name)>
              bar
            </Styled(Name)>
            : 
            <Styled(Type)>
              number
            </Styled(Type)>
          </div>
          <div>
            <Styled(Name)>
              baz
            </Styled(Name)>
            : 
            <Styled(Type)>
              any
            </Styled(Type)>
          </div>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.instanceOf 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        num
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        Number
      </Styled(Type)>
    </div>
    <div />
    <div>
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.objectOf 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        colors
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        {string}
      </Styled(Type)>
    </div>
    <div />
    <div>
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.objectOf(PropTypes.shape) 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        colors
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        {shape}
      </Styled(Type)>
    </div>
    <div />
    <div>
      <div>
        <Styled(Para)>
          <div>
            <Styled(Name)>
              bar
            </Styled(Name)>
            : 
            <Styled(Type)>
              number
            </Styled(Type)>
             — 
            <Styled(Text)>
              Required
            </Styled(Text)>
          </div>
          <div>
            <Styled(Name)>
              baz
            </Styled(Name)>
            : 
            <Styled(Type)>
              any
            </Styled(Type)>
          </div>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.oneOf 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        size
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        enum
      </Styled(Type)>
    </div>
    <div />
    <div>
      <div>
        <Styled(Para)>
          <span>
            One of:
             
            <Group
              inline={true}
              separator=", "
            >
              <Styled(Code)>
                small
              </Styled(Code)>
              <Styled(Code)>
                normal
              </Styled(Code)>
              <Styled(Code)>
                large
              </Styled(Code)>
            </Group>
          </span>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.oneOfType 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        union
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        union
      </Styled(Type)>
    </div>
    <div />
    <div>
      <div>
        <Styled(Para)>
          <span>
            One of type:
             
            <Group
              inline={true}
              separator=", "
            >
              <Styled(Type)>
                string
              </Styled(Type)>
              <Styled(Type)>
                number
              </Styled(Type)>
            </Group>
          </span>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.shape 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        shape
      </Styled(Type)>
    </div>
    <div />
    <div>
      <div>
        <Styled(Para)>
          <div>
            <Styled(Name)>
              bar
            </Styled(Name)>
            : 
            <Styled(Type)>
              number
            </Styled(Type)>
             — 
            <Styled(Text)>
              Required
            </Styled(Text)>
          </div>
          <div>
            <Styled(Name)>
              baz
            </Styled(Name)>
            : 
            <Styled(Type)>
              any
            </Styled(Type)>
          </div>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.shape with description 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        shape
      </Styled(Type)>
    </div>
    <div />
    <div>
      <div>
        <Styled(Para)>
          <div>
            <Styled(Name)>
              bar
            </Styled(Name)>
            : 
            <Styled(Type)>
              number
            </Styled(Type)>
             — 
            <Styled(Text)>
              Required
            </Styled(Text)>
             — 
            <Styled(Markdown)
              inline={true}
              text="Number"
            />
          </div>
          <div>
            <Styled(Name)>
              baz
            </Styled(Name)>
            : 
            <Styled(Type)>
              any
            </Styled(Type)>
             — 
            <Styled(Markdown)
              inline={true}
              text="Any"
            />
          </div>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.shape with formatted defaultProps 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        shape
      </Styled(Type)>
    </div>
    <div>
      <Styled(Text)
        title="{
        bar: 123,
        baz: baz() {
                return 'foo';
            },
        bing: bing() {
                return 'badaboom';
            },
        trotskij: () => 1935,
        qwarc: {
          si: 'señor'
        }
      }"
        underlined={true}
      >
        Shape
      </Styled(Text)>
    </div>
    <div>
      <div>
        <Styled(Para)>
          <div>
            <Styled(Name)>
              bar
            </Styled(Name)>
            : 
            <Styled(Type)>
              number
            </Styled(Type)>
             — 
            <Styled(Text)>
              Required
            </Styled(Text)>
          </div>
          <div>
            <Styled(Name)>
              baz
            </Styled(Name)>
            : 
            <Styled(Type)>
              any
            </Styled(Type)>
          </div>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.string 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        color
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        string
      </Styled(Type)>
    </div>
    <div />
    <div>
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.string with a default value 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        color
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        string
      </Styled(Type)>
    </div>
    <div>
      <Styled(Code)>
        pink
      </Styled(Code)>
    </div>
    <div>
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.string.isRequired 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        color
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        string
      </Styled(Type)>
    </div>
    <div>
      <Styled(Text)>
        Required
      </Styled(Text)>
    </div>
    <div>
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render arguments from JsDoc tags 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        size
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        number
      </Styled(Type)>
    </div>
    <div />
    <div>
      <div>
        <Styled(Markdown)
          text="Test description"
        />
        <JsDoc
          arg={
            Array [
              Object {
                "description": "Converts foo to bar",
                "name": "Foo",
                "type": Object {
                  "name": "Array",
                },
              },
            ]
          }
          param={
            Array [
              Object {
                "name": "Bar",
              },
            ]
          }
        />
        <Styled(Arguments)
          args={
            Array [
              Object {
                "description": "Converts foo to bar",
                "name": "Foo",
                "type": Object {
                  "name": "Array",
                },
              },
              Object {
                "name": "Bar",
              },
            ]
          }
          heading={true}
        />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render description in Markdown 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        color
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        string
      </Styled(Type)>
    </div>
    <div />
    <div>
      <div>
        <Styled(Markdown)
          text="Label"
        />
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render function body in tooltip 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        fn
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        func
      </Styled(Type)>
    </div>
    <div>
      <Styled(Text)
        title="(e) => console.log(e)"
        underlined={true}
      >
        Function
      </Styled(Text)>
    </div>
    <div>
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render name as deprecated when tag deprecated is present 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={true}
      >
        size
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        number
      </Styled(Type)>
    </div>
    <div />
    <div>
      <div>
        <Styled(Markdown)
          text="Test description"
        />
        <JsDoc
          deprecated={
            Array [
              Object {
                "description": "Do not use.",
                "title": "deprecated",
              },
            ]
          }
        />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render unknown proptype for a prop when a relevant proptype is not assigned 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        color
      </Styled(Name)>
    </div>
    <div>
      <Styled(Type)>
        unknown
      </Styled(Type)>
    </div>
    <div>
      <Styled(Code)>
        pink
      </Styled(Code)>
    </div>
    <div>
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;
